Allow pages to be structured as a tree

thehabbos007 opened the pull request:
## 🌟 What is the purpose of this PR?

<!-- Explain, at a high level, what this does and why. -->
<!-- Use the 'What does this change?' section to list more specific implementation details. -->
This PR allows pages to take on a tree structure through links to parent trees.
The API can return a list of pages with references to parents, and through a utility function (`treeFromParentReferences`) this list can be restructured to a tree.

The functionality is exposed in the API by a new mutation, 
`setParentPage(accountId: ID!, pageId: ID!, parentPageId: ID!): Page!`

and a new parameter for the existing resolver for fetching an account's pages
`accountPages(accountId: ID!, structure: PageStructure = Flat): [Page!]!`
The `PageStructure` can be either `Flat` or `Tree` with the difference being that `Tree` makes the `parentPageId` field available.

## ⚠️ Known issues

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->
The main 'issue" is that the tree restructuring has to be done client side because GraphQL is not ergonomic for querying recursive types such as a tree structure. This is why a utility function is exposed for the frontend for UI implementation.

## 🚫 Blocked by

<!-- If the pull request is blocked by anything, list the blockers here. -->
<!-- If applicable, link to them. -->

- [ ] ...

## 🐾 Next steps

<!-- Are there are planned/suggested follow ups which are related but won't be done in this PR? -->
The parent-setting logic could be pulled out of the resolver and allow API users to set a parent when creating a page.

## 🔍 What does this change?

<!-- Use a bullet list to explain your changes in more detail, if it would be helpful. -->
<!-- If applicable, link to the specific commit.-->

- Changes the `accountPages` query to include a (defaulted) argument for fetching references to parent pages
- Adds a new mutation for pages that sets a parent page, checking for cyclic dependencies and only allowing a single parent per page
- Adds a `page.model` model that is used for the interactions regarding exposing parent page references

### Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->

## 🔗 Related links

<!-- Add links to any context it is worth capturing (e.g. Issues, Discussions, Discord, Asana) -->
<!-- Mark any links which are not publically accessible as (_internal_) -->
<!-- Don't rely on links to explain the PR, especially internal ones: use the sections above -->

- [Asana task](https://app.asana.com/0/1201637128042639/1201445315010330/f) (_internal_)

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

- Unit test for utility function
- Integration test for logical error cases and happy path

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->
1. create a 3 pages in the UI
2. Set parents of the pages using their Entity ID to create a tree hierarchy
<details><summary>setParent GQL</summary>

**Query**:

```graphql
mutation setParentPage($accountId: ID!, $pageId: ID!, $parentPageId: ID!) {
  setParentPage(
    accountId: $accountId
    pageId: $pageId
    parentPageId: $parentPageId
  ) {
    entityId
    properties {
      title
      summary
      __typename
    }
    __typename
  }
}
```

**Data**:
```json
{
  "accountId": "change",
  "pageId": "change",
  "parentPageId": "change"
}
```
</details>

3. Check the flat, references page result
<details><summary>accountPages Tree GQL</summary>

**Query**:
```graphql
query getAccountPages($accountId: ID!) {
  accountPages(accountId: $accountId, structure: Tree) {
    entityId
    properties {
      title
      summary
    }
    parentPageId
  }
}
```

**Data**:
```json
{
  "accountId": "change, same as previous query",
}
```
</details>

The `parentPageId` fields of the result should match the initial `setParentPage` created in the first steps.

Error cases can be testes by trying to set the root as the child of another page.

## 📹 Demo

[Demo](https://user-images.githubusercontent.com/6988668/151530121-5ec5a992-ef77-4c1a-99cc-8ba52099dcf4.mp4)

CiaranMn commented on packages/hash/api/jest.config.js:
```diff
@@ -1,5 +1,8 @@
 /** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
 module.exports = {
+  moduleNameMapper: {
+    "@hashintel/hash-shared(.*)": "<rootDir>/../shared/src$1",
+  },
```
What's this for?

thehabbos007 replied:
This is to allow the jest tests to import utilities from the shared project through workspace references. 
With the decision to let the frontend handle restructuring the tree from references, I moved `treeFromParentReferences` to the shared utilities file, the same function is re-exported in the API util file, but the tests are still at the API location

CiaranMn replied:
Understood, thank you

CiaranMn commented on packages/hash/api/src/db/adapter.ts:
```diff
@@ -380,6 +384,16 @@ export interface DBClient {
     systemTypeName: SystemType;
   }): Promise<DbEntity[]>;
 
+  getEntitiesByTypeWithOutgoingEntityIds(params: {
+    accountId: string;
+    systemTypeName: SystemType;
+  }): Promise<EntityWithOutgoingEntityIds[]>;
```
I would either (a) rename this to `getEntitiesBySystemType...` or (b) accept `entityTypeId` as an alternative param to `systemTypeName`

CiaranMn commented on packages/hash/api/src/db/postgres/link/getEntityOutgoingLinks.ts:
```diff
@@ -36,3 +56,59 @@ export const getEntityOutgoingLinks = async (
+group by a.entity_id`;
+
+/**
+ * Get the latest version of all entities of a given type and their outgoing link ids with a matching entity type.
```
It doesn't look like "with a matching entity type" belongs here - it's just getting entities of a given type and their outgoing link ids, no?

thehabbos007 replied:
You're right. I think I removed that part of the query when i switched to a lateral left join. Will add it back, as it would otherwise returns links that might not be relevant for page trees

CiaranMn commented on packages/hash/api/src/db/postgres/link/getEntityOutgoingLinks.ts:
```diff
@@ -36,3 +56,59 @@ export const getEntityOutgoingLinks = async (
+};
+
+/**
+ * Get the latest version of a given entity and all outgoing link ids with a matching entity type.
```
I can't see any matching on entity type going on here

CiaranMn commented on packages/hash/api/src/db/postgres/link/getEntityOutgoingLinks.ts:
```diff
@@ -36,3 +56,59 @@ export const getEntityOutgoingLinks = async (
+ * @param params.accountId the account to retrieve entities from
+ * @param params.entityId entityId of source entity.
+ */
+export const getEntityWithOutgoingEntityIds = async (
```
Is there any potential for consolidating logic/query fragments between this and `getEntityOutgoingLinks`?

CiaranMn commented on packages/hash/api/src/graphql/resolvers/pages/setParentPage.ts:
```diff
@@ -0,0 +1,75 @@
+  MutationSetParentPageArgs
+> = async (_, { accountId, pageId, parentPageId }, { dataSources, user }) => {
+  return await dataSources.db.transaction(async (client) => {
+    const pageEntity = await Page.getAccountPageWithParents(client, {
+      accountId,
+      entityId: pageId,
+    });
+
+    const parentPageEntity = await Entity.getEntityLatestVersion(client, {
+      accountId,
+      entityId: parentPageId,
+    });
```
You could do this in parallel:

```typescript
const [pageEntity, parentPageEntity] = await Promise.all([
  Page.getAccountPageWithParents(...), 
  Entity.getEntityLatestVersion(...)
]);
```

CiaranMn commented on packages/hash/api/src/graphql/resolvers/pages/setParentPage.ts:
```diff
@@ -0,0 +1,75 @@
+        : `'${pageId}' ${
+            parentPageEntity?.entityId
+              ? ""
+              : `nor parent page with entityId = '${parentPageId}`
```
```suggestion
              : ` nor parent page with entityId = '${parentPageId}'`
```

thehabbos007 replied:
https://github.com/hashintel/hash/blob/66b928cec19baa13f8bac0353eb5a545a0ad7d22/packages/hash/api/src/graphql/resolvers/pages/setParentPage.ts#L27 the space is added before the interpolation. But I think i would prefer for it to be here instead, actually. I'll change it :)

CiaranMn replied:
Oh yes, my mistake, apologies

CiaranMn commented on packages/hash/api/src/graphql/resolvers/pages/setParentPage.ts:
```diff
@@ -0,0 +1,75 @@
+      );
+    }
+
+    if (pageEntity.parentEntityId) {
+      throw new Error(
+        `Page with entityId = '${pageEntity.entityId}' already has parent page entityId = '${pageEntity.parentEntityId}'`,
+      );
+    }
```
We should permit changing the parent's page (to support moving pages around)

thehabbos007 replied:
Missed that functionality when implementing - my bad. Will look into that :)

CiaranMn commented on packages/hash/api/src/model/page.model.ts:
```diff
@@ -0,0 +1,42 @@
+} & EntityConstructorArgs;
+
+class __Page extends Entity {
+  parentEntityId?: string;
+  constructor(args: PageConstructorArgs) {
```
```suggestion
  parentEntityId?: string;
  
  constructor(args: PageConstructorArgs) {
```

CiaranMn commented on packages/hash/integration/src/tests/integration.test.ts:
```diff
@@ -1256,6 +1259,157 @@ describe("logged in user ", () => {
+      });
+    });
+
+    it("prevent multiple parents", async () => {
+      // - Super page 1
+      //   - sub page 1             <- because this page already has a parent
+      //     - sub sub page 1
+      //       - sub page 1 <- should not be allowed
+      await expect(
+        client.setParentPage({
+          accountId: existingUser.accountId,
+          pageId: subPage.entityId,
+          parentPageId: subSubPage.entityId,
+        }),
+      ).rejects.toThrowError(
+        /page with entityId = '.*' already has parent page entityId = '.*'/i,
+      );
+    });
```
Have suggested elsewhere that we permit overwriting a page's parent, in which case we can still have this test but check that it's changed to a single parent.

thehabbos007 replied:
This is an interesting test case since re-parenting `sub page 1` under `sub sub page 1` would cause a cyclic dependency. I guess that's another case to solve

CiaranMn commented on packages/hash/shared/src/util.ts:
```diff
@@ -61,3 +62,144 @@ export const isFileProperties = (props: {}): props is FileProperties => {
+/**
+ * Restructure a list of elements with parent references into a tree
+ * This will ignore elements with missing parent entities and add them at the root level
+ * This can be used for recreating a page tree strucutre from the list of an account's pages
```
```suggestion
 * This can be used for recreating a page tree structure from the list of an account's pages
```

CiaranMn reviewed (changes requested):
Looking good - one comment on functionality, a couple of questions and nits.

thehabbos007 commented:
Changes:
- Fixed properly fetching matching types for outgoing links,
- Allow changing the parent (checks for cyclic references)
- Found and fixed a potential issue with deleting links

CiaranMn reviewed (commented):
On links deletion: at the moment, we track which versions of an entity a link is relevant to by checking `source_entity_version_ids` on a link: this array is a list of all the versions of an entity to which the link applies.

That is why, when creating a new link (or deleting or whatever), a new version of the entity is created to be able to add its version_id from the array (or remove it).

I do not like this implementation and would rather we instead stored `linkValidFrom` and `linkValidUntil` timestamps, so that you could just check if a particular entity version fell within that range, rather than having to create new versions alongside updating links and maintain a huge array.

But that's how it works at the moment - so actually the existing delete link behavior is correct. It's just that when you are getting links in your new queries, you are not filtering for only the links that apply to the version of the entity you're dealing with, so when a new link has been created to replace the old you're getting duplicates.

CiaranMn reviewed (approved):
All tested and works well, looks good!
